common/cross-chain/src/deposit_common.rs

```diff
@@ -0,0 +1,133 @@
+use multiversx_sc::storage::StorageKey;
+use operation::{
+    aliases::{ExtractedFeeResult, GasLimit, TxNonce},
+    TransferData,
+};
+use proxies::fee_market_proxy::FeeMarketProxy;
+
+use crate::MAX_TRANSFERS_PER_TX;
+
+multiversx_sc::imports!();
+
+#[multiversx_sc::module]
+pub trait DepositCommonModule:
+    crate::storage::CrossChainStorage + utils::UtilsModule + crate::CrossChainCommon
+{
+    fn match_fee_payment(
+        &self,
+        total_tokens_for_fees: usize,
+        fees_payment: &OptionalValue<EsdtTokenPayment<Self::Api>>,
+        opt_transfer_data: &Option<TransferData<<Self as ContractBase>::Api>>,
+    ) {
+        match fees_payment {
+            OptionalValue::Some(fee) => {
+                let mut gas: GasLimit = 0;
+
+                if let Some(transfer_data) = opt_transfer_data {
+                    gas = transfer_data.gas_limit;
+                }
+
+                let fee_market_address = self.fee_market_address().get();
+                let caller = self.blockchain().get_caller();
+
+                self.tx()
+                    .to(fee_market_address)
+                    .typed(FeeMarketProxy)
+                    .subtract_fee(caller, total_tokens_for_fees, OptionalValue::Some(gas))
+                    .payment(fee.clone())
+                    .sync_call();
+            }
+            OptionalValue::None => (),
+        };
+    }
+
+    fn check_and_extract_fee(&self) -> ExtractedFeeResult<Self::Api> {
+        let payments = self.call_value().all_esdt_transfers().clone();
+
+        require!(!payments.is_empty(), "Nothing to transfer");
+        require!(payments.len() <= MAX_TRANSFERS_PER_TX, "Too many tokens");
+
+        let fee_market_address = self.fee_market_address().get();
+        let fee_enabled_mapper = SingleValueMapper::new_from_address(
+            fee_market_address.clone(),
+            StorageKey::from("feeEnabledFlag"),
+        )
+        .get();
+
+        let fee_payment = if fee_enabled_mapper {
+            OptionalValue::Some(self.pop_first_payment(payments.clone()).0)
+        } else {
+            OptionalValue::None
+        };
+
+        MultiValue2::from((fee_payment, payments))
+    }
+
+    fn burn_mainchain_token(
```

Comment:
This function, some of below #[inline] functions and all mappers multiversx <-> sovereign should be only in `to-sovereign` contract, these are not common because they do not apply for enshine-esdt-safe